test/cpp/lazy/test_trie_cache.cpp

```diff
@@ -0,0 +1,88 @@
+#include <gtest/gtest.h>
+
+#include <c10/util/Exception.h>
+#include <torch/csrc/lazy/core/config.h>
+#include <torch/csrc/lazy/core/ir.h>
+#include <torch/csrc/lazy/core/ir_builder.h>
+#include <torch/csrc/lazy/core/ir_metadata.h>
+#include <torch/csrc/lazy/core/ir_util.h>
+#include <memory>
+
+namespace torch {
+namespace lazy {
+
+class TrieCacheNode : public Node {
+ public:
+  explicit TrieCacheNode(size_t id)
+      : Node(OpKind(), /* num_outputs */ 1), id_(id), hash_(Hash(id_)) {}
+  ~TrieCacheNode() override = default;
+
+  bool Equal(size_t id) const {
+    return (id_ == id);
+  }
+
+  void AddOperand(Value v) {
+    if (!v.node) {
+      return;
+    }
+    operands_as_outputs_.emplace_back(v.node.get(), v.index);
+    operands_.push_back(std::move(v.node));
+  }
+
+  hash_t hash() const override { return hash_; }
+  hash_t shapeHash() const override { return hash_; }
+ private:
+  size_t id_;
+  hash_t hash_;
+};
+
+TEST(TrieCacheTest, TestSinglePath) {
+  FLAGS_torch_lazy_reuse_ir = true;
+  TrieCache::Get()->Clear();
+
+  NodePtr a = MakeNode<TrieCacheNode>(0);
+  NodePtr b = MakeNode<TrieCacheNode>(1);
+  NodePtr c = MakeNode<TrieCacheNode>(2);
+  TrieCache::Get()->ResetCurrent(); // MarkStep
+
+  EXPECT_EQ(ReuseOrMakeNode<TrieCacheNode>(OpKind(), 0).get(), a.get());
```

Comment:
i was surprised you can pass OpKind() here if it is needed for casting, so i looked at NodeCast.  It makes sense now.. no need to change it for this PR.

My idea for cleaning up later is to make opkind a static member of the derived class rather than a field of node base.  Then, could we change NodeCast<T>(node) to access T.kind directly?  (not sure if that is valid in template code)

Reply:
Agree that we need some way to carry the opcode at the class level. But as you pointed out the template code makes it difficult. I will update this once I figure out a good way to do it.

Reply:
Currently the best I can do is to add a static member for each derived class, and make the base Node op() method virtual so that it can return the correct op for each derived class. That will be a fairly big change and definitely worth to be separated in another PR.